site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+  content = re.sub(r'\\slug: [^\s]*', '', content)
+
+  # normalize whitespace
+  content = re.sub(r'\s+', ' ', content)
+  return content
+
+cleaned_documents = []
+
+for d in documents:
+  cleaned_text = clean_up_text(d.text)
+  d.text = cleaned_text
+  cleaned_documents.append(d)
+```
+
+### Embed the preprocessed documents
+
+We'll use Snowflake's Arctic Embed model available from HuggingFace to embed the documents. We'll also use Llama-Index's `SemanticSplitterNodeParser` for processing.
+
+```python
+from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+from llama_index.core.node_parser import SemanticSplitterNodeParser
+
+embed_model = HuggingFaceEmbedding("Snowflake/snowflake-arctic-embed-m")
+
+splitter = SemanticSplitterNodeParser(
+  buffer_size=1, breakpoint_percentile_threshold=85, embed_model=embed_model
+)
+```
+
+With the embed model and splitter, we can execute them in an ingestion pipeline
+
+```python
+from llama_index.core.ingestion import IngestionPipeline
+
+cortex_search_pipeline = IngestionPipeline(
+  transformations=[
+    splitter,
+  ],
+)
+
+results = cortex_search_pipeline.run(show_progress=True, documents=cleaned_documents)
+```
+
+### Load data to Cortex Search
+
+Now that we've embedded our documents, we're ready to load them to Cortex Search.
+
+Here we can use the same connection details as we set up for Cortex Complete.
+
+```python
+import os
+import snowflake.connector
+from tqdm.auto import tqdm
+
+conn = snowflake.connector.connect(
+  user=connection_details["user"],
+  password=connection_details["password"],
+  account=connection_details["account"],
+  warehouse=connection_details["warehouse"],
+  database=connection_details["database"],
+  schema=connection_details["schema"]
+)
+
+conn.cursor().execute("CREATE OR REPLACE TABLE streamlit_docs(doc_text VARCHAR)")
+for curr in tqdm(result):
+  conn.cursor().execute("INSERT INTO streamlit_docs VALUES (%s)", curr.text)
+```
+
+## Calling the Cortex Search Service
+Duration: 5
+
+Here we'll create a CortexSearchRetreiver class to connect to our cortex search service and add the `retrieve` method that we can leverage for calling it.
+
+```python
+import os
+from snowflake.core import Root
+
+class CortexSearchRetriever:
+
+  def __init__(self, session = session, limit_to_retrieve: int = 4):
+    self.session = session
+    self._limit_to_retrieve = limit_to_retrieve
+  
+  def retrieve(self, query: str):
+    root = Root(self.session)
+    cortex_search_service = root.databases[
+        os.environ["SNOWFLAKE_DATABASE"]].schemas[
+          os.environ["SNOWFLAKE_SCHEMA"]].cortex_search_services[
+            os.environ["SNOWFLAKE_CORTEX_SEARCH_SERVICE"]]
+    resp = cortex_search_service.search(
+        query=query,
+        columns=["doc_text"],
+        limit=self._limit_to_retrieve,
+      )
+    if resp.results:
+      return [curr["doc_text"] for curr in resp.results]
+    session.close()
```

Comment:
A few issues here:
1. You're closing `session` not `self.session`.
2. Are you sure you want to close it to begin with? What if you call retrieve multiple times? Given the session was given to this class opened, I don't think it's the responsibility of this class to close it.
3. This won't execute if L251 fires.
4. You should be either throwing an error or returning something here if it fails I presume (perhaps an empty list?)

Reply:
Addressed 1, 3 & 4.

For 2: I do want to close, otherwise it seems to conflict with Complete. This works fine if you retrieve multiple times.